File: grpc/grpc-client-annotation-processor/src/main/java/ru/tinkoff/grpc/client/annotation/processor/GrpcClientExtension.java
package ru.tinkoff.grpc.client.annotation.processor;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.*;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static ru.tinkoff.grpc.client.annotation.processor.GrpcClassNames.*;

public final class GrpcClientExtension implements KoraExtension {
    private final ProcessingEnvironment env;
    private final TypeMirror stubErasure;

    public GrpcClientExtension(ProcessingEnvironment env, TypeMirror stubErasure) {
        this.env = env;
        this.stubErasure = Objects.requireNonNull(stubErasure);
    }

    @Nullable
    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, @Nullable String tag) {
        if (typeMirror.getKind() != TypeKind.DECLARED) {
            return null;
        }
        var dt = (DeclaredType) typeMirror;
        var te = (TypeElement) dt.asElement();
        var className = ClassName.get(te);
        if (env.getTypeUtils().isAssignable(typeMirror, stubErasure)) {
            return getStubGenerator(typeMirror, tag);
        }
        if (className.equals(channel) && tag != null) {
            return getChannel(typeMirror, tag);
        }
        if (className.equals(grpcClientConfig) && tag != null) {
            return getConfig(typeMirror, tag);
        }

        return null;
    }

    private KoraExtensionDependencyGenerator getConfig(TypeMirror typeMirror, String tag) {
        var grpcServiceClassName = ClassName.bestGuess(tag);

        var clientConfigTypeElement = env.getElementUtils().getTypeElement(grpcClientConfig.canonicalName());
        var factoryMethod = findStaticMethod(clientConfigTypeElement, "defaultConfig");
        var parameterTags = new ArrayList<String>(factoryMethod.getParameters().size() - 1);
        var parameterTypes = new ArrayList<TypeMirror>(factoryMethod.getParameters().size() - 1);
        for (int i = 0; i < factoryMethod.getParameters().size() - 1; i++) {
            var parameter = factoryMethod.getParameters().get(i);
            parameterTypes.add(parameter.asType());
            parameterTags.add(null);
        }

        return () -> new ExtensionResult.CodeBlockResult(
            factoryMethod,
            params -> CodeBlock.of("$T.defaultConfig($L, $T.SERVICE_NAME)", grpcClientConfig, params, grpcServiceClassName),
            typeMirror,
            tag,
            parameterTypes,
            parameterTags
        );
    }

    private KoraExtensionDependencyGenerator getChannel(TypeMirror typeMirror, String tag) {
        var grpcServiceTypeElement = env.getElementUtils().getTypeElement(tag);
        var grpcServiceClassName = ClassName.get(grpcServiceTypeElement);
        var managedChannelTypeElement = env.getElementUtils().getTypeElement(managedChannelLifecycle.canonicalName());
        var managedChannelTypeMirror = managedChannelTypeElement.asType();
        var managedChannelConstructor = CommonUtils.findConstructors(managedChannelTypeElement, m -> m.contains(Modifier.PUBLIC)).get(0);
        var parameterTags = new ArrayList<String>(managedChannelConstructor.getParameters().size() - 1);
        var parameterTypes = new ArrayList<TypeMirror>(managedChannelConstructor.getParameters().size() - 1);
        for (int i = 0; i < managedChannelConstructor.getParameters().size() - 1; i++) {
            var parameter = managedChannelConstructor.getParameters().get(i);
            parameterTags.add(i < 4 ? tag : null);
            parameterTypes.add(parameter.asType());
        }
        return () -> new ExtensionResult.CodeBlockResult(
            managedChannelConstructor,
            params -> CodeBlock.of("new $T($L, $T.getServiceDescriptor())", managedChannelLifecycle, params, grpcServiceClassName),
            managedChannelTypeMirror,
            tag,
            parameterTypes,
            parameterTags
        );
    }

    @Nullable
    private KoraExtensionDependencyGenerator getStubGenerator(TypeMirror typeMirror, String tag) {
        if (tag != null) {
            return null;
        }
        if (typeMirror.getKind() != TypeKind.DECLARED) {
            return null;
        }
        var dtm = (DeclaredType) typeMirror;
        var typeElement = dtm.asElement();
        var apiTypeElement = typeElement.getEnclosingElement();
        if (apiTypeElement.getKind() != ElementKind.CLASS) {
            return null;
        }
        var apiClassName = ClassName.get((TypeElement) apiTypeElement);
        if (AnnotationUtils.findAnnotation(apiTypeElement, grpcGenerated) == null) {
            return null;
        }
        var typeName = typeMirror.toString();
        final ExecutableElement sourceElement;
        if (typeName.endsWith("BlockingStub")) {
            sourceElement = this.findStaticMethod(apiTypeElement, "newBlockingStub");
        } else if (typeName.endsWith("FutureStub")) {
            sourceElement = this.findStaticMethod(apiTypeElement, "newFutureStub");
        } else {
            sourceElement = this.findStaticMethod(apiTypeElement, "newStub");
        }
        var channelType = sourceElement.getParameters().get(0).asType();

        return () -> new ExtensionResult.CodeBlockResult(
            sourceElement,
            params -> CodeBlock.of("$T.$N($L)", apiClassName, sourceElement.getSimpleName(), params),
            typeMirror,
            tag,
            List.of(channelType),
            List.of(apiClassName.canonicalName())
        );
    }

    private ExecutableElement findStaticMethod(Element type, String methodName) {
        for (var enclosedElement : type.getEnclosedElements()) {
            if (enclosedElement.getKind() != ElementKind.METHOD) {
                continue;
            }
            if (!enclosedElement.getModifiers().contains(Modifier.STATIC)) {
                continue;
            }
            if (!enclosedElement.getModifiers().contains(Modifier.PUBLIC)) {
                continue;
            }
            if (enclosedElement.getSimpleName().contentEquals(methodName)) {
                return (ExecutableElement) enclosedElement;
            }
        }
        throw new IllegalStateException();
    }
}

File: grpc/grpc-client-symbol-processor/src/main/kotlin/ru.tinkoff.grpc.client.ksp/GrpcClientExtension.kt
package ru.tinkoff.grpc.client.ksp

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.getDeclaredFunctions
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.channel
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.grpcClientConfig
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.grpcGenerated
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.managedChannelLifecycle
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.stubFor
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult.CodeBlockResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValueNoDefault
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.ksp.common.KspCommonUtils.getClassDeclarationByName

class GrpcClientExtension(
    val resolver: Resolver,
    val kspLogger: KSPLogger,
    val codeGenerator: CodeGenerator,
    val abstractStub: KSClassDeclaration?,
    val abstractCoroutineStub: KSClassDeclaration?) : KoraExtension {

    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (abstractStub != null && abstractStub.asStarProjectedType().isAssignableFrom(type)) {
            return this.generateJavaStub(type, tag)
        }
        if (abstractCoroutineStub != null && abstractCoroutineStub.asStarProjectedType().isAssignableFrom(type)) {
            return this.generateCoroutineStub(type, tag)
        }
        val typeName = type.toTypeName()
        if (typeName == channel && tag.size == 1) {
            return this.generateChannel(type, tag)
        }
        if (typeName == grpcClientConfig && tag.size == 1) {
            return this.generateConfig(type, tag)
        }
        return null
    }

    private fun generateConfig(type: KSType, tag: Set<String>): (() -> ExtensionResult) {
        val grpcServiceTypeName = tag.first()
        val grpcServiceClassName = ClassName.bestGuess(grpcServiceTypeName)


        val configClassDecl = resolver.getClassDeclarationByName(grpcClientConfig)!!
        val factoryMethod = findMethod(configClassDecl, "defaultConfig")
        val parameterTags = factoryMethod.parameters.map { setOf<String>() }.dropLast(1)
        val parameterTypes = factoryMethod.parameters.map { it.type.resolve() }.dropLast(1)

        return {
            CodeBlockResult(
                factoryMethod,
                { CodeBlock.of("%T.defaultConfig(%L, %T.SERVICE_NAME)", grpcClientConfig, it, grpcServiceClassName) },
                type,
                tag,
                parameterTypes,
                parameterTags
            )
        }

    }

    private fun findMethod(apiClassDecl: KSClassDeclaration, name: String) = apiClassDecl.getDeclaredFunctions()
        .filter { it.simpleName.asString() == name }
        .first()

    private fun generateChannel(type: KSType, tag: Set<String>): (() -> ExtensionResult) {
        val grpcServiceTypeName = tag.first()
        val grpcServiceClassName = ClassName.bestGuess(grpcServiceTypeName)

        val managedChannelDecl = resolver.getClassDeclarationByName(managedChannelLifecycle.canonicalName)!!
        val managedChannelType = managedChannelDecl.asStarProjectedType()
        val managedChannelConstructor = managedChannelDecl.getConstructors().first()
        val parameterTags = arrayListOf<Set<String>>()
        val parameterTypes = arrayListOf<KSType>()
        for ((i, parameter) in managedChannelConstructor.parameters.dropLast(1).withIndex()) {
            if (i < 3) {
                parameterTags.add(tag)
            } else {
                parameterTags.add(setOf())
            }
            if (i == 1) {
                parameterTypes.add(parameter.type.resolve().makeNullable())
            } else {
                parameterTypes.add(parameter.type.resolve().makeNotNullable())
            }
        }
        return {
            CodeBlockResult(
                managedChannelConstructor,
                { CodeBlock.of("%T(%L, %T.getServiceDescriptor())", managedChannelLifecycle, it, grpcServiceClassName) },
                managedChannelType,
                tag,
                parameterTypes,
                parameterTags
            )
        }
    }

    private fun generateCoroutineStub(type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (tag.isNotEmpty()) {
            return null
        }
        val classDecl = type.declaration
        if (classDecl !is KSClassDeclaration) {
            return null
        }
        val apiClassDecl = classDecl.parentDeclaration
        if (apiClassDecl !is KSClassDeclaration) {
            return null
        }
        val stubForAnnotation = apiClassDecl.findAnnotation(stubFor)
        if (stubForAnnotation == null) {
            return null
        }
        val apiTag = stubForAnnotation.findValueNoDefault<List<KSType>>("value")!!
            .map { it.toClassName().canonicalName }
            .toSet()
        val implClassName = classDecl.toClassName()
        val constructor = classDecl.primaryConstructor!!
        val channelType = constructor.parameters[0].type.resolve()
        return {
            CodeBlockResult(
                classDecl,
                { CodeBlock.of("%T(%L)", implClassName, it) },
                type,
                apiTag,
                listOf(channelType),
                listOf(apiTag)
            )
        }
    }

    private fun generateJavaStub(type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (tag.isNotEmpty()) {
            return null
        }
        val classDecl = type.declaration
        if (classDecl !is KSClassDeclaration) {
            return null
        }
        val apiClassDecl = classDecl.parentDeclaration
        if (apiClassDecl !is KSClassDeclaration) {
            return null
        }
        val apiClassName = apiClassDecl.toClassName()
        if (!apiClassDecl.isAnnotationPresent(grpcGenerated)) {
            return null
        }
        val typeName = classDecl.simpleName.asString()
        val sourceElement = when {
            typeName.endsWith("BlockingStub") -> this.findMethod(apiClassDecl, "newBlockingStub")
            typeName.endsWith("FutureStub") -> this.findMethod(apiClassDecl, "newFutureStub")
            else -> this.findMethod(apiClassDecl, "newStub")
        }
        val channelType = sourceElement.parameters[0].type.resolve()
        return {
            CodeBlockResult(
                sourceElement,
                { CodeBlock.of("%T.%N(%L)", apiClassName, sourceElement.simpleName.asString(), it) },
                type,
                tag,
                listOf(channelType),
                listOf(setOf(apiClassName.canonicalName))
            )
        }
    }
}

